### PR TITLE
[GradedAxes] Introduce LabelledUnitRangeDual

### PR DESCRIPTION
**Pull request by ogauthe**
This PR aims to fix cases where dual information is lost while slicing a `BlockSparseArray`. There are 2 cases:
- slicing a `GradedUnitRangeDual` with a `Block` now produces a (non-blocked) `LabelledUnitRangeDual`, which preserves dual information
- slicing a `GradedUnitRangeDual` with a vector of `Block` or `BlockIndexRange` produces a `BlockVector` whose only axis is a `GradedUnitRangeDual`. This is needed in views of `BlockSparseArray`.

### CODE REVIEW

File: NDTensors/src/lib/GradedAxes/src/gradedunitrangedual.jl [hunk 1]
```diff
@@ -49,20 +53,50 @@ function BlockArrays.blocklengths(a::GradedUnitRangeDual)
...
+  # GradedOneTo appears in mortar
+  # flip v axis to preserve dual information
+  # axes(v) will appear in axes(view(::BlockSparseArray, [Block(1)]))
+  return flip_blockvector(v)
+end
+
+function flip_blockvector(v::BlockVector)
+  # TODO way to create BlockArray with specified axis without relying on internal?
+  block_axes = flip.(axes(v))
+  flipped = BlockArrays._BlockArray(vec.(blocks(v)), block_axes)
```
Comment (mtfishman) [inline NDTensors/src/lib/GradedAxes/src/gradedunitrangedual.jl:98]: I think you can use `mortar(vec.(blocks(v)), block_axes)`.

File: NDTensors/src/lib/GradedAxes/src/gradedunitrangedual.jl [hunk 2]
```diff
@@ -49,20 +53,50 @@ function BlockArrays.blocklengths(a::GradedUnitRangeDual)
   return dual.(blocklengths(nondual(a)))
 end
 
-function gradedunitrangedual_getindices_blocks(a::GradedUnitRangeDual, indices)
+# TODO: Move this to a `BlockArraysExtensions` library.
+function blockedunitrange_getindices(
+  a::GradedUnitRangeDual, indices::Vector{<:BlockIndexRange{1}}
+)
   a_indices = getindex(nondual(a), indices)
   return mortar([label_dual(b) for b in blocks(a_indices)])
```
Comment (mtfishman) [inline NDTensors/src/lib/GradedAxes/src/gradedunitrangedual.jl:61]: Should this output a `GradedUnitRangeDual`?

Comment (mtfishman) [reply NDTensors/src/lib/GradedAxes/src/gradedunitrangedual.jl:61]: Never mind, I guess it isn't a unit range.

**Comment by mtfishman** [comment]
Looks good to me, ready to merge from your end?

**Comment by ogauthe** [comment]
With this PR, we now have an error with
```julia
label(dual(labelled(1:1, "x")))
```
```julia
ERROR: MethodError: no method matching dual(::String)

Closest candidates are:
  dual(::NDTensors.SymmetrySectors.Ising)
   @ NDTensors ~/Documents/itensor/ITensors.jl/NDTensors/src/lib/SymmetrySectors/src/sector_definitions/ising.jl:24
  dual(::NDTensors.SymmetrySectors.Fib)
   @ NDTensors ~/Documents/itensor/ITensors.jl/NDTensors/src/lib/SymmetrySectors/src/sector_definitions/fib.jl:24
  dual(::NDTensors.SymmetrySectors.TrivialSector)
   @ NDTensors ~/Documents/itensor/ITensors.jl/NDTensors/src/lib/SymmetrySectors/src/sector_definitions/trivial.jl:15
  ...

Stacktrace:
 [1] label(a::LabelledUnitRangeDual{LabelledInteger{Int64, String}, LabelledUnitRange{Int64, UnitRange{Int64}, String}})
   @ NDTensors.GradedAxes ~/Documents/itensor/ITensors.jl/NDTensors/src/lib/GradedAxes/src/labelledunitrangedual.jl:16
 [2] top-level scope
   @ REPL[12]:1
   ```

Maybe we should default `dual(x) = x`.

**Comment by mtfishman** [comment]
> Maybe we should default `dual(x) = x`.

That would make sense as a definition, we do that as a default definition for unit ranges anyway.

File: NDTensors/src/lib/GradedAxes/src/gradedunitrange.jl [hunk 3]
```diff
@@ -68,6 +68,7 @@ end
 # == is just a range comparison that ignores labels. Need dedicated function to check equality.
 struct NoLabel end
 blocklabels(r::AbstractUnitRange) = Fill(NoLabel(), blocklength(r))
+blocklabels(la::LabelledUnitRange) = label(la)
```
Comment (mtfishman) [inline NDTensors/src/lib/GradedAxes/src/gradedunitrange.jl:71]:
```suggestion
blocklabels(la::LabelledUnitRange) = [label(la)]
```
`blocklabels(r)` should output a collection of labels of length `blocklength(r)`.

File: NDTensors/src/lib/GradedAxes/src/labelledunitrangedual.jl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+       AbstractUnitRange{T}
+  nondual_unitrange::NondualUnitRange
+end
+
+dual(a::LabelledUnitRange) = LabelledUnitRangeDual(a)
+nondual(a::LabelledUnitRangeDual) = a.nondual_unitrange
+dual(a::LabelledUnitRangeDual) = nondual(a)
+flip(a::LabelledUnitRangeDual) = dual(flip(nondual(a)))
+isdual(::LabelledUnitRangeDual) = true
+blocklabels(la::LabelledUnitRangeDual) = label(la)
```
Comment (mtfishman) [inline NDTensors/src/lib/GradedAxes/src/labelledunitrangedual.jl:15]:
```suggestion
blocklabels(la::LabelledUnitRangeDual) = [label(la)]
```

**Comment by ogauthe** [comment]
Ready to merge!

**Comment by mtfishman** [comment]
> Ready to merge!

Great, thanks. I'll merge once tests pass.

**Pull request merged**